bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -228,28 +235,84 @@ boolean doRefresh() {
+   *          case of an {@link IOException}, create a new {@link BackOff}; see
+   *          {@link RetryOptions#createBackoff()} for more details on the initial construction. If
+   *          a retry already occurred, the Backoff have the logic to get the next sleep period and
+   *          retry exhaustion logic.
+   * @return HeaderCacheElement containing either a valid {@link AccessToken} or an exception.
+   */
+  protected HeaderCacheElement refreshCredentialsWithRetry(@Nullable BackOff backoff) {
     try {
       AccessToken newToken = credentials.refreshAccessToken();
       if (newToken == null) {
-        // The current implementations of refreshAccessToken() throw an IOException or return
-        // a valid token. This handling is future proofing.
-        cacheElement = new HeaderCacheElement(
-            new IOException("Could not load the token for credentials: " + credentials));
+        // The current implementations of refreshAccessToken() throws an IOException or return
+        // a valid token. This handling is future proofing just in case credentials returns null for
+        // some reason. This case was caught by a poorly coded unit test.
+        return new HeaderCacheElement(new IOException("Could not load the token for credentials: "
+            + credentials));
       } else {
-        cacheElement = new HeaderCacheElement(newToken);
+        // Success!
+        return new HeaderCacheElement(newToken);
+      }
+    } catch (IOException exception) {
+      // An IOException occurred. Retry with backoff.
+      if (backoff == null) {
+        // initialize backoff.
+        backoff = retryOptions.createBackoff();
+      }
+
+      // Given the backoff, either sleep for a short duration, or terminate if the backoff has
+      // reached its configured timeout limit.
+      RetryState retryState = getRetryState(backoff);
+      if (retryState == RetryState.PerformRetry) {
+        return refreshCredentialsWithRetry(backoff);
+      } else {
+        return new HeaderCacheElement(exception);
       }
-    } catch (IOException e) {
-      cacheElement = new HeaderCacheElement(e);
     } catch (Exception e) {
-      cacheElement = new HeaderCacheElement(new IOException("Could not read headers", e));
-    } finally {
-      synchronized (isRefreshing) {
-        headerCache.set(cacheElement);
-        isRefreshing.set(false);
-        isRefreshing.notifyAll();
+      return new HeaderCacheElement(new IOException("Could not read headers", e));
+    }
+
+  }
+
+  /**
+   * Sleep and/or determine if the backoff has timed out.
+   * @param backoff
+   * @return RetryState indicating the current state of the retry logic.
+   */
+  protected RetryState getRetryState(BackOff backoff) {
+    long nextBackOffMillis = 0;
+    try {
+      nextBackOffMillis = backoff.nextBackOffMillis();
+      if (nextBackOffMillis == BackOff.STOP) {
+        return RetryState.RetriesExhausted;
```

Comment:
I think we can log a message when are get the response from getRetryState(backoff);